sandbox/apparmor/prompting/message.go

```diff
@@ -0,0 +1,424 @@
+	if err := binary.Write(msgBuf, order, raw); err != nil {
+		return nil, err
+	}
+	if _, err := msgBuf.Write(packer.Bytes()); err != nil {
+		return nil, err
+	}
+	return msgBuf.Bytes(), nil
+}
+
+// Validate returns an error if the mssage contains invalid data.
+func (msg *MsgNotificationFilter) Validate() error {
+	if !msg.ModeSet.IsValid() {
+		return fmt.Errorf("unsupported modeset: %d", msg.ModeSet)
+	}
+	return nil
+}
+
+// MsgNotification describes a kernel notification message.
+//
+// This structure corresponds to the kernel type struct apparmor_notif
+// described below.
+//
+//	struct apparmor_notif {
+//	  struct apparmor_notif_common base;
+//	  __u16 ntype;        /* notify type */
+//	  __u8 signalled;
+//	  __u8 flags;
+//	  __u64 id;           /* unique id, not globally unique*/
+//	  __s32 error;        /* error if unchanged */
+//	} __attribute__((packed));
+type MsgNotification struct {
+	MsgHeader
+	// NotificationType describes the kind of notification message used.
+	// Currently the kernel only sends APPARMOR_NOTIF_OP messages.
+	NotificationType NotificationType
+	// XXX: Signaled seems to be unused.
+	Signalled uint8
+	// Set to 1 to NOT cache
+	Flags uint8
+	// ID is an opaque kernel identifier of the notification message. It must be
+	// repeated in the MsgNotificationResponse if one is sent back.
+	ID uint64
+	// XXX: This seems to be unused and clashes with identical field in
+	// MsgNotificationOp.
+	Error int32
+}
+
+// UnmarshalBinary unmarshals the message from binary form.
+func (msg *MsgNotification) UnmarshalBinary(data []byte) error {
+	const prefix = "cannot unmarshal apparmor notification message"
+
+	// Unpack fixed-size elements.
+	buf := bytes.NewBuffer(data)
+	order := arch.Endian() // ioctl messages are native byte order, verify endianness if using for other messages
+	if err := binary.Read(buf, order, msg); err != nil {
+		return fmt.Errorf("%s: cannot unpack: %s", prefix, err)
+	}
+
+	return nil
+}
+
+// MarshalBinary marshals the message into binary form.
+func (msg *MsgNotification) MarshalBinary() ([]byte, error) {
+	msg.Version = 2
+	msg.Length = uint16(binary.Size(*msg))
+	buf := bytes.NewBuffer(make([]byte, 0, msg.Length))
+	order := arch.Endian() // ioctl messages are native byte order, verify endianness if using for other messages
+	if err := binary.Write(buf, order, msg); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+// Validate returns an error if the mssage contains invalid data.
+func (msg *MsgNotification) Validate() error {
+	if !msg.NotificationType.IsValid() {
+		return fmt.Errorf("unsupported notification type: %d", msg.NotificationType)
+	}
+	return nil
+}
+
+// MsgNotificationUpdate (TBD, document me)
+//
+//	struct apparmor_notif_update {
+//	  struct apparmor_notif base;
+//	  __u16 ttl;          /* max keep alives left */
+//	} __attribute__((packed));
+type MsgNotificationUpdate struct {
```

Comment:
Do we need this type?

Reply:
Really good catch, thanks. I don't see it used anywhere, either in @mvo5 's [cerberus-dbus](https://github.com/mvo5/snappy/tree/cerberus-dbus) branch or in the newer WIP [prompting](https://github.com/olivercalder/snapd/tree/prompting) branch. I'll check with Michael once he's back, but I don't know what it was/is meant to be used for. I think it can likely be removed, and re-added later if we decide we want TTL functionality (whatever that means in this context).

Reply:
@jrjohansen should this type of message be used for anything? I removed it, but can re-add it if it's useful.

Reply:
it is part of the hear beat/keep alive mechanism. snapd is not using it atm

Reply:
To answer in this context, it means registering your notification to do this. You get a heart beat message every X seconds, that you need to reply to so the message doesn't time out. This allows for a notification to not have a hard fixed time out (2 minutes atm), so if snapd hangs apparmor will wake the application sooner.

It is something we will probably want to use in the future, but not at the moment.

Reply:
Thanks for the explanation! I'll leave it removed, and we can re-add it and discuss how to handle it when we want to use it in the future.